### PR TITLE
Design Picker: Group results by the best matching and categories

### PR DESCRIPTION
**Pull request by arthur791004**
Related to https://github.com/Automattic/dotcom-forge/issues/10095

## Proposed Changes

* Group results by the best matching themes, and then each selected category
* Rename some copy (replace `design` with `theme`)

image
image

**Notes**
* It looks weird if the best matching themes are the same as themes for a single category
* There may be lots of themes for a category so people need to scroll down a lot to see the themes for the next category

## Why are these changes being made?

*

## Testing Instructions

* Create a new site
* On the Goals screen, select any goals
* On the Design Picker screen, select any topic & types
* Ensure the result looks good
* Select `Community & Non-Profit` & enable the `Free only` toggle
* Ensure the `No themes matched.` result looks good
* Add the `flags=-design-picker/goal-centric` to the end of the URL
* Ensure the Design Picker screen on production still looks good

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: packages/design-picker/src/components/unified-design-picker.tsx [hunk 1]
```diff
@@ -266,26 +352,32 @@ const DesignPicker: React.FC< DesignPickerProps > = ( {
...
-							getBadge={ getBadge }
-							oldHighResImageLoading={ oldHighResImageLoading }
-							isActive={ showActiveThemeBadge && design.recipe?.stylesheet === siteActiveTheme }
-						/>
-					);
-				} ) }
-				{ filteredDesigns.length === 0 && <NoResults /> }
-			</div>
+			{ isMultiFilterEnabled && (
+				<DesignCardGroup
```
Comment (arthur791004) [inline packages/design-picker/src/components/unified-design-picker.tsx:361]: do we need to limit the best matches to at least 2 selected categories?

Comment (taipeicoder) [reply packages/design-picker/src/components/unified-design-picker.tsx:361]: Yes please.

**Review by taipeicoder — APPROVED** [review]
LGTM 👍 

One thing I noticed is that newly selected filters are appended to the bottom instead of putting it at the top of the results. This is not ideal as it makes users have to scroll to see what the newly selected filter offers.

**Review by fushar — COMMENTED** [review]
> One thing I noticed is that newly selected filters are appended to the bottom instead of putting it at the top of the results. This is not ideal as it makes users have to scroll to see what the newly selected filter offers.

As a dev who's still catching up with this project, I didn't realize this behavior until I saw this comment. I mean, I didn't realize that when I click multiple filters, the new results are appended to the bottom. I thought nothing happened because simply there are no better  matching sets of themes to show.

Not sure how to "fix" this, though 🤔

File: packages/design-picker/src/components/unified-design-picker.tsx [hunk 2]
```diff
@@ -268,26 +354,32 @@ const DesignPicker: React.FC< DesignPickerProps > = ( {
...
+			{ Object.entries( designsByGroup ).map(
+				( [ categorySlug, categoryDesigns ], index, array ) => (
+					<DesignCardGroup
+						key={ categorySlug }
+						{ ...designCardProps }
+						title={
+							isMultiFilterEnabled
+								? `${ ( categorization?.categories || [] ).find(
+										( { slug } ) => slug === categorySlug
+								  )?.name } themes`
```
Comment (fushar) [inline packages/design-picker/src/components/unified-design-picker.tsx]: Hmm, should this be translated?

Comment (arthur791004) [reply packages/design-picker/src/components/unified-design-picker.tsx]: Ok

**Comment by arthur791004** [comment]
> Is that because we don't have any overlap?

Yes

> Another one is when I have the blog filter, there's a weird scrolling thing that throws me back to the top of the screen:

Fixed by [81a95d2](https://github.com/Automattic/wp-calypso/pull/97263/commits/81a95d2a2ccbce4b606995fcfe71c4f332aba496).

> I thought best matching would be the overlap between two filters. If we have a single filter, we shouldn't show best matching

Sorry for not being clear. I mean themes under a category may all have the same categories, so the overlap between two is still the same as themes under that category. For example, the result of the Blog & Newsletter is the same as the Newsletter designs.

**Review by taipeicoder — APPROVED** [review]
Looks awesome to me 🚀

**Comment by arthur791004** [comment]
Since there were failing e2e tests, reverted it and opened https://github.com/Automattic/wp-calypso/pull/97314 again to address the issue

**Pull request merged**